src/Database/LSMTree/Internal/Run.hs

```diff
@@ -141,3 +151,63 @@ fromWriteBuffer fs fsPaths buffer = do
     for_ (WB.content buffer) $ \(k, e) ->
       addFullKOp fs mrun k e
     fromMutable fs mrun
+
+data ChecksumError = ChecksumError FS.FsPath CRC.CRC32C CRC.CRC32C
+  deriving Show
+
+instance Exception ChecksumError
+
+data FileFormatError = FileFormatError FS.FsPath String
+  deriving Show
+
+instance Exception FileFormatError
+
+-- | Load a previously written run from disk, checking each file's checksum
+-- against the checksum file.
+openFromDisk :: HasFS IO h -> RunFsPaths -> IO (Run (FS.Handle h))
+openFromDisk fs lsmRunFsPaths = do
+    expectedChecksums <-
+       expectValidFile (runChecksumsPath lsmRunFsPaths) . fromChecksumsFile
+         =<< CRC.readChecksumsFile fs (runChecksumsPath lsmRunFsPaths)
+
+    -- verify checksums of files we don't read yet
+    let paths = runFsPaths lsmRunFsPaths
+    checkCRC (forRunKOps expectedChecksums) (forRunKOps paths)
+    checkCRC (forRunBlob expectedChecksums) (forRunBlob paths)
+
+    -- read and try parsing files
+    lsmRunFilter <-
+      expectValidFile (forRunFilter paths) . bloomFilterFromSBS
+        =<< readCRC (forRunFilter expectedChecksums) (forRunFilter paths)
+    (lsmRunNumEntries, lsmRunIndex) <-
+      expectValidFile (forRunIndex paths) . Index.fromSBS
+        =<< readCRC (forRunIndex expectedChecksums) (forRunIndex paths)
+
+    lsmRunKOpsFile <- FS.hOpen fs (runKOpsPath lsmRunFsPaths) FS.ReadMode
+    lsmRunBlobFile <- FS.hOpen fs (runBlobPath lsmRunFsPaths) FS.ReadMode
+    lsmRunRefCount <- newIORef 1
+    return Run {..}
+  where
+    checkCRC :: CRC.CRC32C -> FS.FsPath -> IO ()
+    checkCRC expected fp = do
+        checksum <- CRC.readFileCRC32C fs fp
+        expectChecksum fp expected checksum
+
+    readCRC :: CRC.CRC32C -> FS.FsPath -> IO SBS.ShortByteString
+    readCRC expected fp = FS.withFile fs fp FS.ReadMode $ \h -> do
+        -- Read the whole file, which should usually return a single chunk.
+        -- In this case, 'toStrict' is O(1).
+        size <- FS.hGetSize fs h
+        (lbs, crc) <- CRC.hGetExactlyCRC32C fs h size CRC.initialCRC32C
+        expectChecksum fp expected crc
+        let bs = LBS.toStrict lbs
+        return $ case tryCheapToShort bs of
+          Right sbs -> sbs
+          Left _err -> SBS.toShort bs  -- Should not happen, but just in case
+
+    expectChecksum fp expected checksum =
+        when (expected /= checksum) $
+          throw $ ChecksumError fp expected checksum
+
+    expectValidFile _  (Right x)  = pure x
+    expectValidFile fp (Left err) = throw $ FileFormatError fp err
```

Comment:
A case could be made to use `throwIO` instead of `throw`, because it guarantees ordering of exceptions

Reply:
A convincing case, if you ask me. :+1: